### ISSUE TITLE
mlContentField insertion order leading to serious data loss

### ISSUE THREAD

**Issue by herrvigg** [herrvigg/qtranslate-x#70]
<a href="https://github.com/funkjedi"><img src="https://avatars2.githubusercontent.com/u/9314?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [funkjedi](https://github.com/funkjedi)**
_Thursday Mar 05, 2015 at 23:50 GMT_
_Originally opened as https://github.com/qTranslate-Team/qtranslate-x/issues/70_

----

When qTranslate-X adds content hooks `qtranxj_ce()` inserts the `mlContentField` at the top of the form. This results in the fields being added in reverse order. Under normal circumstances this has no adverse affects.

However it can have serious affects when the field names utilize the array notation. For example:

``` html
<input name="acf[field][0][field_2]" type="hidden" class="hidden" value="">
<input name="acf[field][0][field_1]" type="hidden" class="hidden" value="">
<input name="acf[field]" type="hidden" class="hidden" value="">
```

As you can see the third field in the DOM essentially wipes out the values of the other fields when the form is loaded into the `$_POST` array by PHP causing some serious data loss.

I've worked around it for now by hooking `qtranxj_ce()` and manually reordering the DOM nodes. It's a pretty ugly hack but it works for now.

``` js
(function($) {

    // save reference to function and then
    // replace with a hooked version
    var _qtranxj_ce = qtranxj_ce;
    qtranxj_ce = function(tagName, props, pNode, isFirst) {

        // call original function
        var el = _qtranxj_ce(tagName, props, pNode, isFirst);

        // detach ACF related mlContentField inputs from DOM
        var form = $(pNode);
        var fields = form.children('input[name^=acf]:hidden').detach().toArray();

        // sort mlContentField inputs by input[name]
        fields.sort(function (a,b) {
            if (a.name > b.name) {
                return 1;
            }
            if (a.name < b.name) {
                return -1;
            }
            return 0;
        });

        // insert ACF related mlContentField inputs back into DOM
        form.prepend(fields);

        // return element created by original function
        return el;

    };

})(jQuery);
```

**Comment by herrvigg** [comment]
<a href="https://github.com/johnclause"><img src="https://avatars2.githubusercontent.com/u/5491361?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [johnclause](https://github.com/johnclause)**
_Friday Mar 06, 2015 at 20:35 GMT_

----

Tim, there is a compatibility problem with "Visual Composer" and other "Visual Whatever": https://github.com/qTranslate-Team/qtranslate-x/issues/27. It seems to me the current design of mlContentField related things will have to be changed soon to enable better compatibility. We may not rename fields anymore in future. That is why it is better not to rely on the low level specifics of the code and only use high level interface. We can change the high level interface to accommodate your needs.

On this particular issue, I never knew that order actually matters, although it is kind of making sense, after you mentioned it. What if we always append to the end of form? This will keep the order. Can you think of anything which may get broken if we do that?

**Comment by herrvigg** [comment]
<a href="https://github.com/funkjedi"><img src="https://avatars2.githubusercontent.com/u/9314?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [funkjedi](https://github.com/funkjedi)**
_Saturday Mar 07, 2015 at 00:04 GMT_

----

I think always appending to the end of the form would work great. The fields would still be direct descendants of the form so I can't imagine anything would break.

**Comment by herrvigg** [comment]
<a href="https://github.com/johnclause"><img src="https://avatars2.githubusercontent.com/u/5491361?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [johnclause](https://github.com/johnclause)**
_Saturday Mar 07, 2015 at 04:37 GMT_

----

I also do not see what can break, so the latest on github puts it at the end. Is there still anything you need to change on -X?

**Comment by herrvigg** [comment]
<a href="https://github.com/johnclause"><img src="https://avatars2.githubusercontent.com/u/5491361?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [johnclause](https://github.com/johnclause)**
_Tuesday Mar 10, 2015 at 04:41 GMT_

----

Hi Tim, I am about to release next update based on the latest at github. Are you all right? Can you do all you need?

**Comment by herrvigg** [comment]
<a href="https://github.com/johnclause"><img src="https://avatars2.githubusercontent.com/u/5491361?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [johnclause](https://github.com/johnclause)**
_Tuesday Mar 10, 2015 at 19:24 GMT_

----

I assume it is closed. Please, open new one if there is more to worry about.

**Comment by herrvigg** [comment]
<a href="https://github.com/johnclause"><img src="https://avatars2.githubusercontent.com/u/5491361?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [johnclause](https://github.com/johnclause)**
_Tuesday Mar 31, 2015 at 18:19 GMT_

----

Hi Tim (@funkjedi), I cannot believe it, but putting fields to the end of form gave trouble to one of the users: https://wordpress.org/support/topic/problems-with-exisiting-posts. If you are in the mood to solve puzzles, I can give you access to the user test site. Indeed, the post does not get saved, when "content" input field is the last one in the form. But it only a trouble for some posts, and not the other. Can you imagine that? I did not dig it into the WP code, there must be some explanation.

For now, I changed common.js again to insert hidden fields right before their originals. This should also keep the order for you. Could you test please, the latest on github?

**Comment by herrvigg** [comment]
<a href="https://github.com/johnclause"><img src="https://avatars2.githubusercontent.com/u/5491361?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [johnclause](https://github.com/johnclause)**
_Wednesday Apr 01, 2015 at 22:27 GMT_

----

Tim, I think the order would be fine for you, what I more worry about is that those hidden input fields in the middle will add some spaces on the view, although, in theory, they should not. You have pretty complex pages, and I hope that you would notice some layout changes. Would you find time to test it before I break a few thousands of users ;)
